### DISCUSSION TITLE
[ Plugin Dev ] [ Javascript ] Unable to save plugin settings after enter is pressed by the user

### DISCUSSION THREAD

**Discussion by gyaaniguy** [Flow-Launcher/Flow.Launcher#2995] [Q&A]
Writing a plugin that's shows a periodical reminder.  I do this by using Javascript's built in `setInterval`  
This works, but I also need to save the return value of the above function call, so it can be cancelled.

So I tried saving the value to plugin settings file. 
However I am only able to write to the settings file when the results are generated. But the setting needs to be updated *after* the user chooses the right option and presses enter.

Below works. But I don't want the setting to update when the user types.
```javascript 

if (method === 'query') {
  console.log(JSON.stringify(
        { 
              result: [getResRow(title, subTitle, durationInMs)],
              SettingsChange: {
                 "reminder_period": "20 minutes"
             }
        })
    );
```

Each row has an action 
```javascript
res['JsonRPCAction'] = {
            method: 'run_remind_loop',
            parameters: [durationInMs, subTitle],
        };

// Handle enter ?
if (method === 'run_remind_loop') {
   // LOOP RUNS !
    let milliseconds = parameters[0];
    run_remind_loop(milliseconds);
    // BELOW HAS NO EFFECT
   console.log(JSON.stringify({
              result: [],
              SettingsChange: {
                  "reminder_period": val
              }
          })
      );
}
```

Any ideas on how to make this work ? Also I notice sometimes pressing enter has no effect on the window. The code works, but the windows stays..

**Comment by gyaaniguy** [comment]
bump !

**Comment by taooceros** [comment]
~~You can pass a `SettingsChanges` inside the result.~~

~~https://github.com/Flow-Launcher/Flow.Launcher/blob/943425bfa16515c9581d8807cdfdf7ca0f3fa4da/Flow.Launcher.Core/Plugin/JsonPRCModel.cs#L56-L62~~

Sorry the above is wrong, as you would need to pre-register the settings change before executing it.

**Reply by Yusyuriv** [reply]
I think this is correct for the problem of saving the settings after the user selects a result. Modify your `getResRow` to also accept the new settings value:

```js
function getResRow(title, subtitle, durationInMs, settingsChange) {
  return {
    ..., // your normal stuff
    SettingsChange: settingsChange
  }
}
```

Now you can use it like this:

```js
console.log(JSON.stringify({ 
  result: [getResRow(title, subTitle, durationInMs, { reminder_period: "20 minutes" })]
}));
```

However, there is another problem: API v1 plugins are supposed to launch, respond with the results, and exit. Every time the user types a character, a new plugin process launches. You want it to persist so it could show periodic notifications. API v2 is much better suited for that, it launches just one instance of the plugin, and that instance persists until you exit Flow Launcher, but unfortunately it's not documented at the moment.

**Reply by gyaaniguy** [reply]
Thanks @Yusyuriv I had totally forgotten about this. I'll try this tomorrow.
